miner/proposal_builder.go

```diff
@@ -317,141 +366,174 @@ func (pb *ProposalBuilder) decideMeshHash(ctx context.Context, current types.Lay
-		pb.session.eligibilities.proofs = calcEligibilityProofs(
-			pb.signer.VRFSigner(),
-			pb.session.epoch,
-			pb.session.beacon,
-			pb.session.nonce,
-			pb.session.eligibilities.slots,
+	if ss.session.eligibilities.proofs == nil {
+		ss.session.eligibilities.proofs = calcEligibilityProofs(
+			ss.signer.VRFSigner(),
+			ss.session.epoch,
+			ss.session.beacon,
+			ss.session.nonce,
+			ss.session.eligibilities.slots,
 			pb.cfg.layersPerEpoch,
 		)
-		pb.logger.With().Info("proposal eligibilities for an epoch", log.Inline(pb.session))
+		ss.log.With().Info("proposal eligibilities for an epoch", log.Inline(&ss.session))
 		events.EmitEligibilities(
-			pb.session.epoch,
-			pb.session.beacon,
-			pb.session.atx,
-			uint32(len(pb.session.active.set)),
-			pb.session.eligibilities.proofs,
+			ss.session.epoch,
+			ss.session.beacon,
+			ss.session.atx,
+			uint32(len(pb.shared.active.set)),
+			ss.session.eligibilities.proofs,
 		)
 	}
 	return nil
 }
 
 func (pb *ProposalBuilder) build(ctx context.Context, lid types.LayerID) error {
-	latency := latencyTracker{start: time.Now()}
-	if err := pb.initSessionData(ctx, lid); err != nil {
+	start := time.Now()
+	if err := pb.initSharedData(ctx, lid); err != nil {
 		return err
 	}
-	latency.data = time.Now()
 
-	if lid <= pb.session.prev {
-		return fmt.Errorf("layer %d was already built", lid)
+	pb.mu.Lock()
+	// don't accept registration in the middle of computing proposals
+	signers := maps.Values(pb.signers)
+	pb.mu.Unlock()
+
+	var eg errgroup.Group
+	eg.SetLimit(pb.cfg.workersLimit)
+	for _, ss := range signers {
+		ss := ss
+		ss.latency.start = start
+		eg.Go(func() error {
+			if err := pb.initSignerData(ctx, ss, lid); err != nil {
+				if errors.Is(err, errAtxNotAvailable) {
+					ss.log.With().Info("smesher doesn't have atx that targets this epoch",
+						log.Context(ctx), log.Uint32("epoch", ss.session.epoch.Uint32()),
+					)
+				} else {
+					return err
+				}
+			}
+			if lid <= ss.session.prev {
+				return fmt.Errorf(
+					"layer %d was already built by signer %s",
+					lid,
+					ss.signer.NodeID().ShortString(),
+				)
+			}
+			ss.session.prev = lid
+			ss.latency.data = time.Now()
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return err
 	}
-	pb.session.prev = lid
 
-	proofs := pb.session.eligibilities.proofs[lid]
-	if len(proofs) == 0 {
-		pb.logger.With().Debug("not eligible for proposal in layer",
-			log.Context(ctx), log.Uint32("lid", lid.Uint32()),
-			log.Uint32("epoch", lid.GetEpoch().Uint32()),
-		)
+	any := false
+	for _, ss := range signers {
+		if n := len(ss.session.eligibilities.proofs[lid]); n == 0 {
+			ss.log.With().Debug("not eligible for proposal in layer",
+				log.Context(ctx),
+				log.Uint32("lid", lid.Uint32()), log.Uint32("epoch", lid.GetEpoch().Uint32()))
+			continue
+		} else {
+			ss.log.With().Debug("eligible for proposals in layer",
+				log.Context(ctx),
+				log.Uint32("lid", lid.Uint32()), log.Int("num proposals", n),
```

Comment:
types.LayerID implements `Field()` already (also in other places):
```suggestion
				lid, log.Int("num proposals", n),
```

Reply:
those fields should be deprecated. they can't be passed as is to zap. and because of that every log method needs to copy whole collection and "unpack" fields

Reply:
i understood what you meant, changed